### PR TITLE
Handle gestures to start edition on the calendar

### LINKED ISSUE #362: Handle tap to start editing new item
## Description
This needs to add a long press gesture to the calendar view and when that happens on an empty space, send the EmptyPositionLongPressed action.

## Dependencies
Depends on #360 
Depends on #361 

## Definition of done
- [x] Get the time for the empty position and send the action

### LINKED ISSUE #363: Handle tap on calendar item
## Description
This needs to add a tap gesture to the calendar view and when that happens on a calendar item, send the ItemSelected action.

## Dependencies
Depends on #360
Depends on #361

## Definition of done
- [x] Get the item at the tapped position and send the action

### PR DESCRIPTION
**Pull request by daividssilverio**
### Summary
- This PR ~simply~ makes the calendar selector consider the selected item from the state, and;
- It handles taps on calendar items;
- It handles long presses on empty spaces;
- It changes `CalendarItem` to include the `SelectedItem` in it, updating the calendar layout calculator and drawing delegate accordingly.

### Technical considerations
~I think the logic is bit convoluted here to be honest...
But I'm not really sure how to improve it... All the `SelectedCalendarItem` <-> `CalendarItem` <-> `EditableTimeEntry` <-> `TimeEntry` logic there makes me uneasy...~

I've updated the `CalendarItem` sealed class to look more like the iOS version.
Now it includes a `SelectedItem` in it, which removes some the paint I've scratched above.

It also simplified the calendar drawing data logic 👍 

### Why is it done like this and not some other way?
Feel free to make suggestions. 
~I'm not really happy with what I wrote but I'm not really sure how to to improve it...~
I've changed it to look like the iOS version, which seems good to me. But again, feel free to make suggestions.

### Relationships
Closes #265
Closes #362 
Closes #363 

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests
- [ ] One of our [live or file custom templates](https://github.com/toggl/android/tree/master/.idea/templates) was used in this PR

### Pointers
- Check out the tests.
- Run and test the calendar.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
~I'm leaving this open as PR for now for 🐎~
~But I'm using this branch to work on the touch related issues for the calendar. I was planning to have a single PR, but this might work best.~

Nobody reviewed this, so I've updated it to close the other two PRs and the `CalendarItem` refactor.

🦑 it

### CODE REVIEW

File: calendar/src/main/java/com/toggl/calendar/calendarday/domain/CalendarLayoutCalculator.kt [hunk 1]
```diff
@@ -67,8 +68,12 @@ class CalendarLayoutCalculator @Inject constructor(private val timeService: Time
      * @return An list of calendar attributes
      */
     private fun calculateLayoutAttributesFor(bucket: CalendarItemGroup): List<CalendarItem> {
-        val left = bucket.filter { it is CalendarItem.CalendarEvent }.toMutableList()
-        val right = bucket.filter { it is CalendarItem.TimeEntry }.toMutableList()
+        val left = bucket.filter {
+            it is CalendarItem.CalendarEvent || it is CalendarItem.SelectedItem && it.selectedCalendarItem is SelectedCalendarItem.SelectedCalendarEvent
+        }.toMutableList()
```
Comment (semanticer) [inline calendar/src/main/java/com/toggl/calendar/calendarday/domain/CalendarLayoutCalculator.kt:73]: You know me I would change those group types to read-only lists and get rid of these mutable operations
```kotlin
typealias CalendarItemGroup = List<CalendarItem>
typealias CalendarItemGroups = List<CalendarItemGroup>
```

File: calendar/src/main/java/com/toggl/calendar/calendarday/ui/views/CalendarWidgetView.kt [hunk 2]
```diff
@@ -444,6 +459,30 @@ class CalendarWidgetView @JvmOverloads constructor(
...
+        val nextDay = currentDate.plusDays(1)
+
+        val offset = currentDate + duration
+
+        if (offset < currentDate)
+            return currentDate.atOffset(currentOffset)
+        if (offset > nextDay)
+            return nextDay.atOffset(currentOffset)
+
+        return offset.atOffset(currentOffset)
```
Comment (semanticer) [inline calendar/src/main/java/com/toggl/calendar/calendarday/ui/views/CalendarWidgetView.kt:485]:
```suggestion
        return when {
            offset < currentDate -> currentDate.atOffset(currentOffset)
            offset > nextDay -> nextDay.atOffset(currentOffset)
            else -> offset.atOffset(currentOffset)
        }
```

File: calendar/src/main/java/com/toggl/calendar/calendarday/ui/views/CalendarWidgetView.kt [hunk 3]
```diff
@@ -527,6 +573,22 @@ class CalendarWidgetView @JvmOverloads constructor(
...
+                return it.selectedCalendarItemToDraw
+        }
+
+        for (item in it.nonSelectedCalendarItemsToDraw) {
+            calendarItemDrawingDelegate.calculateItemRect(item, viewFrame, touchRectF)
+            if (touchRectF.contains(x, y + scrollOffset))
+                return item
+        }
+
+        null
```
Comment (semanticer) [inline calendar/src/main/java/com/toggl/calendar/calendarday/ui/views/CalendarWidgetView.kt:589]:
```suggestion
        it.nonSelectedCalendarItemsToDraw.find { calendarItem ->
            calendarItemDrawingDelegate.calculateItemRect(calendarItem, viewFrame, touchRectF)
            touchRectF.contains(x, y + scrollOffset)
        }
```

File: calendar/src/main/java/com/toggl/calendar/calendarday/domain/CalendarLayoutCalculator.kt (hunk 1 above)
Comment (daividssilverio) [reply calendar/src/main/java/com/toggl/calendar/calendarday/domain/CalendarLayoutCalculator.kt:73]: I got it to work, but I'm not really sure if we should change that, more specifically, I don't think we should make that change in this PR.
I'm gonna open a PR just for that after this branch is merged and we can discuss to keep it or not, how does that sound to you?

Comment (semanticer) [reply calendar/src/main/java/com/toggl/calendar/calendarday/domain/CalendarLayoutCalculator.kt:73]: Sure, sounds good to me

**Pull request merged**